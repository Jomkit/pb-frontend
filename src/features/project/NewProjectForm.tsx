import * as Yup from 'yup';
import DynamicForm from '../../components/DynamicForm';
import userContext from '../../components/contexts/userContext';
import { useContext } from 'react';
import Api from '../../api';
import { useNavigate } from 'react-router-dom';

const newProjectSchema = Yup.object().shape({
    name: Yup.string().required(),
    note: Yup.string().notRequired(),
})

const NewProjectForm = () => {
    const initialProjectValues = {
        name: '',
        note: '',
    }

    const currUser = useContext(userContext);
    const navigate = useNavigate();
    
    const handleSubmit = (values: {name: string, note: string}) => {
        try{
            console.log("Submitting values:", values);
            Api.createUserProject(currUser.id!, values);
            navigate("/projects");
        }catch(err: any){
            console.error("Error detected in NewProjectForm:", (err[0] || err));
            alert(err[0] || err);
            navigate("/projects/new");
        }
    }
    
  return (
    <div className='m-8'>
        <h2>Create New Project</h2>
        <DynamicForm initialValues={initialProjectValues} schema={newProjectSchema} handleSubmit={handleSubmit} />
    </div>
  )
}

export default NewProjectForm