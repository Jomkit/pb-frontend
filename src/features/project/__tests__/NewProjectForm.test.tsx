import { render } from "@testing-library/react"
import NewProjectForm from "../NewProjectForm"
import userEvent from "@testing-library/user-event";
import { MemoryRouter, useNavigate } from "react-router-dom";

vi.mock("../../../api");
it('should render without crashing', () => { 
    render(
        <MemoryRouter>
            <NewProjectForm />
        </MemoryRouter>
    );
})

it('should match snapshot', () => { 
    const { asFragment } = render(
        <MemoryRouter>
            <NewProjectForm />
        </MemoryRouter>
    )
    expect(asFragment()).toMatchSnapshot();
})

it("should redirect to same page when successfully submitting form", async () => {
    const user = userEvent.setup();
    vi.mock("react-router-dom", async (importOriginal) => {
        const mod = await importOriginal<typeof import("react-router-dom")>();
        return {
            ...mod,
            useNavigate: vi.fn(() => `redirected`)
        };
    })

    const { getByLabelText, getByText } = render(
        <MemoryRouter>
            <NewProjectForm />
        </MemoryRouter>
    );

    const submitBtn = getByText("Submit");

    const nameInput = getByLabelText("name");
    const noteInput = getByLabelText("note");

    await user.type(nameInput, "testProject");
    await user.type(noteInput, "testNote");
    await user.click(submitBtn);
    expect(useNavigate).toHaveBeenCalled();
})