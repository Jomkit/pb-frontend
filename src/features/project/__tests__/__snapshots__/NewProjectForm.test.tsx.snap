// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="m-8"
  >
    <h2>
      Create New Project
    </h2>
    <form
      action="#"
    >
      <div
        class="rounded-lg border border-gray-200 bg-white shadow-md dark:border-gray-700 dark:bg-gray-800 flex flex-col text-start w-1/2 m-auto"
        data-testid="flowbite-card"
      >
        <div
          class="flex h-full flex-col justify-center gap-4 p-6"
        >
          <label
            class="block text-gray-700 text-sm font-bold mb-2"
            for="name"
          >
            name
          </label>
          <input
            class="form-field1 focus:outline-none focus:shadow-outline"
            id="name"
            name="name"
            type="text"
            value=""
          />
          <label
            class="block text-gray-700 text-sm font-bold mb-2"
            for="note"
          >
            note
          </label>
          <input
            class="form-field1 focus:outline-none focus:shadow-outline"
            id="note"
            name="note"
            type="text"
            value=""
          />
        </div>
      </div>
      <button
        class="btn bg-blue-700 text-white"
        type="submit"
      >
        Submit
      </button>
    </form>
  </div>
</DocumentFragment>
`;
